File: fc-nodejs/FCstring.js
import { config } from "dotenv";
import { BASE_URL, OPENAI_API_KEY, MODEL_NAME} from "./config.js";
import { OpenAI } from "openai";
config();

const openai = new OpenAI({
    apiKey: OPENAI_API_KEY, 
    baseURL: BASE_URL,
});

const count_words = async(text)=>{
    if (typeof text !== 'string') {
        text = String(text);
    }
    return text.split(/\s+/).length;
}

const find_substring = async(text, substring)=>{
    const positions = [];
    let startIndex=0;
    while(true){
        const index = text.indexOf(substring, startIndex);
        if(index===-1){
            break;
        }
        positions.push(index);
        startIndex=index+substring.length;
    }
    return positions;
}

const extract_emails = async(text)=>{
    const pattern = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;
    return text.match(pattern);

}

const word_frequency = async(text)=>{
    if (typeof text !== 'string') {
        text = String(text);
    }
    const words = text.split(/\s+/);
    const frequency = {};
    for(const word of words){
        if(word.length>0){
            frequency[word]=(frequency[word]||0)+1
        }
    }
    return frequency;
}

const text_statistics = async(text)=>{
    if (typeof text !== 'string') {
        text = String(text);
    }
    const words = text.split(/\s+/);
    const sentences = text.split(/[.!?]+/).filter(sentence => sentence.trim().length > 0);
    const word_count = words.length;
    const sentence_count = sentences.length;
    
    const sum = (arr) => arr.reduce((a, b) => a + b, 0);
    
    const average_word_length = word_count > 0 ? sum(words.map(word => word.length)) / word_count : 0;
    const average_sentence_length = sentence_count > 0 ? sum(sentences.map(sentence => sentence.length)) / sentence_count : 0;
    
    return {
        word_count,
        sentence_count,
        average_word_length,
        average_sentence_length,
    }
}

const format_text = async(text, max_length) => {
    if (typeof text !== 'string') {
        text = String(text);
    }
    const words = text.split(/\s+/);
    const lines = [];
    let current_line = "";
    let current_length = 0;
    
    for(const word of words){
        if(current_line.length + word.length + 1 <= max_length){
            current_line += (current_line ? " " : "") + word;
            current_length = current_line.length;
        }
        else{
            if(current_line) {
                lines.push(current_line);
            }
            current_line = word;
            current_length = word.length;
        }
    }
    if(current_line){
        lines.push(current_line);
    }
    return lines.join("\n");
}

const executeFunctionCall = async(func_name, args) => {
    switch(func_name){
        case "count_words":
            return await count_words(args.text);
        case "find_substring":
            return await find_substring(args.text, args.substring);
        case "extract_emails":
            return await extract_emails(args.text);
        case "word_frequency":
            return await word_frequency(args.text);
        case "text_statistics":
            return await text_statistics(args.text);
        case "format_text":
            return await format_text(args.text, args.max_length);
        default:
            console.error(`Unknown function name: ${func_name}`);
            return null;
    }
}

const tools = [{
    type: "function",
    name:"count_words",
    description:"count the number of words in a given text",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to count the words for"
            }
        },
        required:["text"]
    }
},
{
    type:"function",
    name:"find_substring",
    description:"find the positions of a substring in a given text",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to find the substring in"
            },
            substring:{
                type:"string",
                description:"the substring to find in the given text"
            }
        },
        required:["text", "substring"]
    }
},
{
    type:"function",
    name:"extract_emails",
    description:"extract all the email addresses from a given text",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to extract the emails from"
            }
        },
        required:["text"]
    }
},
{
    type:"function",
    name: "word_frequency",
    description:"calculate the frequency of each word in a given text",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to calculate the word frequency for"
            }
        },
        required: ["text"]
    }
},
{
    type:"function",
    name:"text_statistics",
    description:"calculate some statistics about a given text",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to calculate the statistics for"
            }
        },
        required:["text"]
    }
},
{
    type:"function",
    name:"format_text",
    description:"format a given text to a given max length",
    parameters:{
        type:"object",
        properties:{
            text:{
                type:"string",
                description:"the given text to format"
            },
            max_length:{
                type:"number",
                description:"the max length to format the text to"
            }
        },
        required:["text", "max_length"]
    }
}]



export const function_call_playground_string = async(prompt)=>{
   console.log("function call playground for string starting...");
   const messages = [
    {
        role:"system",
        content: "You are a string assistant. You can count the number of words, find given substrings, extract emails, calculate word frequency, calculate text statistics, and format text based on given max length, all for a given text. When prompted with one of these tasks, respond with a JSON object containing the type and value. For example: {\"type\": \"count_words\", \"value\": \"Hello, world!\"}. For find_substring, use format: {\"type\": \"find_substring\", \"value\": \"Hello, world!\", \"substring\": \"world\"}. For word frequency, use format: {\"type\": \"word_frequency\", \"value\": \"Hello, world!\"}. For text statistics, use format: {\"type\": \"text_statistics\", \"value\": \"Hello, world!\"}. For format_text, use format: {\"type\": \"format_text\", \"value\": \"Hello, world!\", \"max_length\": 10}"
    },
    {
        role:"user",
        content:prompt
    }
   ]

   try{
    const response = await openai.chat.completions.create({
        model:MODEL_NAME,
        messages:messages,
        temperature: 0.01,
        stream: false,
        top_p: 0.95,
        response_format: { type: "json_object" }
    })
    
    const responseContent = JSON.parse(response.choices[0].message.content);
    console.log("API Response:", responseContent);

    let validationResult;
    switch(responseContent.type){
        case "count_words":
            validationResult = await count_words(responseContent.value);
            break;
        case "find_substring":
            validationResult = await find_substring(responseContent.value, responseContent.substring);
            break;
        case "extract_emails":
            validationResult = await extract_emails(responseContent.value);
            break;
        case "word_frequency":
            validationResult = await word_frequency(responseContent.value);
            break;
        case "text_statistics":
            validationResult = await text_statistics(responseContent.value);
            break;
        case "format_text":
            validationResult = await format_text(responseContent.value, responseContent.max_length);
            break;
        default:
            throw new Error(`Unknown validation type: ${responseContent.type}`);
    }

    return JSON.stringify({
        type:responseContent.type,
        value:responseContent.value,
        isValid:validationResult
    })
   }
   catch(error){
    console.error("API Error:", {
        status:error.status,
        message:error.message,
        type: error.type,
        headers: error.headers,
        response: error.response?.data
    })
   }
}

const prompt = [
    "count the number of words in the text 'Hello, world!'",
    "find the positions of the substring 'world' in the text 'Hello, world!'",
    "extract all the email addresses from the text 'Hello, world! my email is john.doe@example.com and jane.smith@example.com'",
    "calculate the frequency of each word in the text 'Hello, world! my email is john.doe@example.com and jane.smith@example.com'",
    "calculate the statistics of the text 'Hello, world! my email is john.doe@example.com and jane.smith@example.com'",
    "format the text 'Hello, world! my email is john.doe@example.com and jane.smith@example.com' to a max length of 10",
]

for(const p of prompt){
    console.log("Prompt:", p);
    const start= Date.now();
    const result = await function_call_playground_string(p);
    const end=Date.now();
    console.log("Result:", result);
    console.log(`Time taken: ${end-start/1000}seconds`);
    console.log("--------------------------------");
}

File: function-calling/fc-python/FCstring.py
import re
import os
import json
import time
from openai import OpenAI
import sys
import requests
from config import OPENAI_API_KEY, BASE_URL, MODEL_NAME

client = OpenAI(
    api_key=OPENAI_API_KEY,
    base_url=BASE_URL
)
print("client started")

def count_words(text: str) -> int:
    """Count the number of words in a text."""
    return len(text.split())

def find_substring(text: str, substring: str) -> dict:
    """Find all occurrences of a substring in a text."""
    positions = [m.start() for m in re.finditer(re.escape(substring), text)]
    return {
        "count": len(positions),
        "positions": positions
    }

def extract_emails(text: str) -> list:
    """Extract all email addresses from a text."""
    email_pattern = r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
    return re.findall(email_pattern, text)

def word_frequency(text: str) -> dict:
    """Calculate word frequency in a text."""
    words = text.lower().split()
    frequency = {}
    for word in words:
        word = re.sub(r'[^\w\s]', '', word)
        if word:
            frequency[word] = frequency.get(word, 0) + 1
    return frequency

def text_statistics(text: str) -> dict:
    """Calculate various text statistics."""
    words = text.split()
    sentences = re.split(r'[.!?]+', text)
    sentences = [s.strip() for s in sentences if s.strip()]
    
    return {
        "word_count": len(words),
        "sentence_count": len(sentences),
        "avg_word_length": sum(len(word) for word in words) / len(words) if words else 0,
        "avg_sentence_length": len(words) / len(sentences) if sentences else 0
    }

def format_text(text: str, max_length: int = 80) -> str:
    """Format text to fit within a maximum line length."""
    words = text.split()
    lines = []
    current_line = []
    current_length = 0
    
    for word in words:
        if current_length + len(word) + 1 <= max_length:
            current_line.append(word)
            current_length += len(word) + 1
        else:
            lines.append(' '.join(current_line))
            current_line = [word]
            current_length = len(word)
    
    if current_line:
        lines.append(' '.join(current_line))
    
    return '\n'.join(lines)

tools = [{
    "type": "function",
    "function": {
        "name": "count_words",
        "description": "Count the number of words in a text",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to count words in"
                }
            },
            "required": ["text"],
            "additionalProperties": False
        }
    }
}, {
    "type": "function",
    "function": {
        "name": "find_substring",
        "description": "Find all occurrences of a substring in a text",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to search in"
                },
                "substring": {
                    "type": "string",
                    "description": "The substring to find"
                }
            },
            "required": ["text", "substring"],
            "additionalProperties": False
        }
    }
}, {
    "type": "function",
    "function": {
        "name": "extract_emails",
        "description": "Extract all email addresses from a text",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to extract emails from"
                }
            },
            "required": ["text"],
            "additionalProperties": False
        }
    }
}, {
    "type": "function",
    "function": {
        "name": "word_frequency",
        "description": "Calculate word frequency in a text",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to analyze"
                }
            },
            "required": ["text"],
            "additionalProperties": False
        }
    }
}, {
    "type": "function",
    "function": {
        "name": "text_statistics",
        "description": "Calculate various text statistics",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to analyze"
                }
            },
            "required": ["text"],
            "additionalProperties": False
        }
    }
}, {
    "type": "function",
    "function": {
        "name": "format_text",
        "description": "Format text to fit within a maximum line length",
        "parameters": {
            "type": "object",
            "properties": {
                "text": {
                    "type": "string",
                    "description": "The text to format"
                },
                "max_length": {
                    "type": "integer",
                    "description": "Maximum line length (default: 80)"
                }
            },
            "required": ["text"],
            "additionalProperties": False
        }
    }
}]

def function_call_playground(prompt: str):
    messages = [
        {"role": "system", "content": prompt}
    ]

    response = client.chat.completions.create(
        model=MODEL_NAME,
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )

    func1_name = response.choices[0].message.tool_calls[0].function.name
    func1_args = response.choices[0].message.tool_calls[0].function.arguments
    func1_out = eval(f'{func1_name}(**{func1_args})')

    messages.append(response.choices[0].message)
    messages.append({
        'role': 'tool',
        'content': f'{func1_out}',
        'tool_call_id': response.choices[0].message.tool_calls[0].id
    })

    response = client.chat.completions.create(
        model=MODEL_NAME,
        messages=messages,
        temperature=0.01,
        stream=False,
        top_p=0.95,
        tools=tools
    )

    return response.choices[0].message.content

# Test prompts
prompts = [
    "Count the words in this text: Hello world, this is a test message.",
    "Find all occurrences of 'the' in this text: The quick brown fox jumps over the lazy dog.",
    "Extract email addresses from this text: Contact us at test@example.com or support@company.com",
    "Calculate word frequency in this text: The cat and the dog and the cat and the mouse",
    "Get text statistics for this text: This is a test. It has multiple sentences! How many words?",
    "Format this text to 40 characters per line: This is a long text that needs to be formatted to fit within a specific line length while maintaining readability."
]

if __name__ == "__main__":
    for prompt in prompts:
        start_time = time.time()
        print(f"\nPrompt: {prompt}")
        print(f"Response: {function_call_playground(prompt)}")
        end_time = time.time()
        print(f"Time taken: {end_time - start_time:.2f} seconds") 